src/lib.rs

```diff
@@ -469,6 +555,129 @@ pub fn get_current_username() -> Option<String> {
+}
+
+/// Return the group ID for the user running the process.
+pub fn get_current_gid() -> gid_t {
+    OSUsers::empty_cache().get_current_gid()
+}
+
+/// Return the groupname of the user running the process.
+pub fn get_current_groupname() -> Option<String> {
+    OSUsers::empty_cache().get_current_groupname()
+}
+
+/// Return the group ID for the effective user running the process.
+pub fn get_effective_gid() -> gid_t {
+    OSUsers::empty_cache().get_effective_gid()
+}
+
+/// Return the groupname of the effective user running the process.
+pub fn get_effective_groupname() -> Option<String> {
+    OSUsers::empty_cache().get_effective_groupname()
+}
+
+/// Set current user for the running process, requires root priviledges.
+pub fn set_current_uid(uid: uid_t) -> Result<(), io::Error> {
+    match unsafe { setuid(uid) } {
+        0 => Ok(()),
+        -1 => Err(io::Error::last_os_error()),
+        _ => unreachable!()
+    }
+}
+
+/// Set current group for the running process, requires root priviledges.
+pub fn set_current_gid(gid: gid_t) -> Result<(), io::Error> {
+    match unsafe { setgid(gid) } {
+        0 => Ok(()),
+        -1 => Err(io::Error::last_os_error()),
+        _ => unreachable!()
+    }
+}
+
+/// Set effective user for the running process, requires root priviledges.
+pub fn set_effective_uid(uid: uid_t) -> Result<(), io::Error> {
+    match unsafe { seteuid(uid) } {
+        0 => Ok(()),
+        -1 => Err(io::Error::last_os_error()),
+        _ => unreachable!()
+    }
+}
+
+/// Set effective user for the running process, requires root priviledges.
+pub fn set_effective_gid(gid: gid_t) -> Result<(), io::Error> {
+    match unsafe { setegid(gid) } {
+        0 => Ok(()),
+        -1 => Err(io::Error::last_os_error()),
+        _ => unreachable!()
+    }
+}
+
+/// Atomically set current and effective user for the running process, requires root priviledges.
+pub fn set_both_uid(ruid: uid_t, euid: uid_t) -> Result<(), io::Error> {
+    match unsafe { setreuid(ruid, euid) } {
+        0 => Ok(()),
+        -1 => Err(io::Error::last_os_error()),
+        _ => unreachable!()
+    }
+}
+
+/// Atomically set current and effective group for the running process, requires root priviledges.
+pub fn set_both_gid(rgid: gid_t, egid: gid_t) -> Result<(), io::Error> {
+    match unsafe { setregid(rgid, egid) } {
+        0 => Ok(()),
+        -1 => Err(io::Error::last_os_error()),
+        _ => unreachable!()
+    }
+}
+
+pub struct SwitchUserGuard {
+    uid: uid_t,
+    gid: gid_t,
+}
+
+impl Drop for SwitchUserGuard {
+    fn drop(&mut self) {
+        // Panic on error here, as failing to set values back
+        // is a possible security breach.
+        set_effective_uid(self.uid).unwrap();
+        set_effective_gid(self.gid).unwrap();
+    }
+}
+
+/// Safely switch user and group for the current scope.
+/// Requires root access.
+///
+/// ```rust
+/// {
+///     let guard = switch_user_group(1001, 1001);
```

Comment:
Actually, if you don't 'ignore', you should at least 'no_run' this test - otherwise cargo will actually try to run it, and unless you're running cargo as root (bad idea) it'll still fail. (I think. I haven't actually done this...)